azure/services/networkinterfaces/networkinterfaces.go

```diff
@@ -116,7 +129,40 @@ func (s *Service) Delete(ctx context.Context) error {
 	return result
 }
 
-// IsManaged returns always returns true as CAPZ does not support BYO network interfaces.
+// isManaged returns true if the network interface has an owned tag with the cluster name as value,
+// meaning that the network interface's lifecycle is managed.
+func (s *Service) isManaged(ctx context.Context, spec azure.ResourceSpecGetter) (bool, error) {
+	_, _, done := tele.StartSpanWithLogger(ctx, "networkinterfaces.Service.IsManaged")
+	defer done()
+
+	if spec == nil {
+		return false, errors.New("cannot get network interface to check if it is managed: spec is nil")
+	}
+
+	scope := azure.NetworkInterfaceID(s.Scope.SubscriptionID(), spec.ResourceGroupName(), spec.ResourceName())
+	if s.TagsGetter == nil {
+		return false, fmt.Errorf("tagsgetter is nil")
+	}
+	if ctx == nil {
+		return false, fmt.Errorf("ctx is nil")
+	}
+	result, err := s.TagsGetter.GetAtScope(ctx, scope)
+	if err != nil {
+		return false, err
+	}
+
+	tagsMap := make(map[string]*string)
+	if result.Properties != nil && result.Properties.Tags != nil {
+		tagsMap = result.Properties.Tags
+	}
+
+	tags := converters.MapToTags(tagsMap)
+	return tags.HasOwned(s.Scope.ClusterName()), nil
+}
+
+// IsManaged returns true if the network interface has an owned tag with the cluster name as value,
+// meaning that the network interface's lifecycle is managed.
 func (s *Service) IsManaged(ctx context.Context) (bool, error) {
+	// This method is unused, but it is required to satisfy the ServiceReconciler interface
```

Comment:
I may be misunderstanding something here, but would it make sense to move the contents of the unexported `isManaged` here in case the method needs to be used elsewhere? Otherwise, I can see it being an issue in the future if someone needs to use the exported `IsManaged` and gets a true return value even if the NIC is not managed.

Reply:
This is indeed confusing. The problem is that the interface this is fulfilling is external. It is called during reconciling delete of an AzureCluster. https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/4cfde3590d3f8b4db78031c64508c72c03b8e342/controllers/azurecluster_reconciler.go#L111
I'd argue that check isn't necessary. The check could be done in the services' `Delete()` implementations. And by stubbing out the call to always return true, the `Delete()` implementation in the networkinterfaces service can decide on a per-interface basis whether it should delete any network interfaces.

Reply:
Actually, to clarify further, that call isn't even made for the networkinterfaces service, only the resource groups service. The stubbed method is unused, but must be present on all the services to ensure they fulfill the interface.

Reply:
The `IsManaged()` function was meant to be eventually used to determine if a resource is BYO (and we should skip reconcile) or if its lifecycle is managed by CAPZ. There are several services that implement it and use it (vnets, subnets, etc.), although not as part of the interface.

Since we're adding support for BYO NIC we should be able to use it to return `false` if a NIC is pre-existing. 

Not sure how much of this is impacted by https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3608 though, @nojnhuh PTAL

Reply:
The problem is that this function can only be called per AzureMachine, not per network interface. Since an AzureMachine can have multiple interfaces, it is possible that some are BYO and others are not. Unless we choose to enforce that you must either bring all your own NIC's or none of them, this function won't be usable as it is.

Reply:
I don't think #3608 or ASO in general will fix this particular issue, I think it's something we have to account for in either case. I think the current approach of inlining individual NIC-specific `isManaged()` checks and letting the interface `IsManaged()` no-op seems reasonable for now, but I agree it's worth looking for ways to generalize this a bit more in the future.

Reply:
^ Actually I think the current ASO framework will handle this case since it doesn't use the service-level `IsManaged()` and instead will decide whether or not to skip reconciling individual resources based on whether they're managed or not, just like how it's done inline here.